source/extensions/filters/http/admission_control/admission_control.h

```diff
@@ -0,0 +1,123 @@
+#pragma once
+
+#include <cstdint>
+#include <memory>
+#include <string>
+
+#include "envoy/common/time.h"
+#include "envoy/extensions/filters/http/admission_control/v3alpha/admission_control.pb.h"
+#include "envoy/http/codes.h"
+#include "envoy/http/filter.h"
+#include "envoy/runtime/runtime.h"
+#include "envoy/server/filter_config.h"
+#include "envoy/stats/scope.h"
+#include "envoy/stats/stats_macros.h"
+
+#include "common/common/cleanup.h"
+#include "common/grpc/common.h"
+#include "common/grpc/status.h"
+#include "common/http/codes.h"
+#include "common/runtime/runtime_protos.h"
+
+#include "extensions/filters/http/admission_control/evaluators/response_evaluator.h"
+#include "extensions/filters/http/admission_control/thread_local_controller.h"
+#include "extensions/filters/http/common/pass_through_filter.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace AdmissionControl {
+
+/**
+ * All stats for the admission control filter.
+ */
+#define ALL_ADMISSION_CONTROL_STATS(COUNTER) COUNTER(rq_rejected)
+
+/**
+ * Wrapper struct for admission control filter stats. @see stats_macros.h
+ */
+struct AdmissionControlStats {
+  ALL_ADMISSION_CONTROL_STATS(GENERATE_COUNTER_STRUCT)
+};
+
+using AdmissionControlProto =
+    envoy::extensions::filters::http::admission_control::v3alpha::AdmissionControl;
+
+/**
+ * Configuration for the admission control filter.
+ */
+class AdmissionControlFilterConfig {
+public:
+  AdmissionControlFilterConfig(const AdmissionControlProto& proto_config, Runtime::Loader& runtime,
+                               TimeSource& time_source, Runtime::RandomGenerator& random,
+                               Stats::Scope& scope, ThreadLocal::SlotPtr&& tls,
+                               std::unique_ptr<ResponseEvaluator> response_evaluator);
+  virtual ~AdmissionControlFilterConfig() = default;
+
+  virtual ThreadLocalController& getController() const {
+    return tls_->getTyped<ThreadLocalControllerImpl>();
```

Comment:
I don't remember if/when we discussed this, but what's the tradeoff for a global controller vs. per worker? Would it be better to go with global for better accuracy similar to what we do for adaptive concurrency?

Reply:
Adaptive concurrency required global control because of the way it took measurements. We needed high levels of accuracy because of the percentile calculations- we would be in danger of missing tail latencies. We also needed to globally limit the concurrency across all worker threads since it would no longer mean that limiting concurrency (on one thread) resulted in decreased load on the local host. It would almost be the same as using adaptive concurrency on the egress. There's also complications around synchronization of the measurement windows, etc. Global just made more sense.

For this filter, the measurements aren't as finicky and making it thread-local would simply just decrease the number of samples. This is fine, since there's not really a control loop that other threads can interfere with (not sure if I'm using that correctly, but I mean we're not making a change and then measuring its effect). The decision-making here can be decentralized and we still get roughly the same results.

The benefits to thread-local controllers is obviously that we don't need to take any locks on the deque/structures that store the sampling history. The drawbacks are that it's a bit more complicated to implement and we need a copy of the sampling history deque/structures for each thread. I decided that the memory overhead is minimal, so it's worth it to make it thread-local. In cases where memory utilization might raise an eyebrow, we'd be at high risk for lock contention.

Reply:
That's fine. Add some comments.

Reply:
I think one potential drawback with TLS is that incoming traffic isn't necessarily evenly distributed between the workers, so it could result in cases where overall we're rejecting some % of traffic, but individual clients that are connected to workers that haven't seen much traffic might be able to pass through without any rejections.

You could imagine 3 clients: client 1 and 2 are connected to worker A and client 3 to worker B. Client 1 starts sending traffic thats failing, triggering throttling. Client 2 and 3 are both sending identical traffic, but only client 2 sees throttling because it happened to be handled by the worker that is also handling client 1.

That said I think TLS is fine assuming we document the behavior.